docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,177 @@
+  unblock these APIs.
+* Place for staging experimental APIs: New APIs can remain in seperate
+  federated servers until they become stable, at which point, they can be moved
+  to the core kubernetes master, if appropriate.
+
+## Goal
+
+* Developers should be able to write their own API server and cluster admins
+  should be able to add them to their cluster, exposing new APIs at runtime. All
+  of this should not require any change to the core kubernetes API server.
+* These new APIs should be seamless extension of the core kubernetes APIs (ex:
+  they should be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+There will be a reverse proxy in front of all API servers. Clients will talk to
+the proxy to discover the API groups and resources exposed by all of them.
+Clients can use the proxy to talk to federated servers, without having to know
+their individual URLs.
+
+### Constraints
+
+* Unique API groups across servers: Each API server should expose unique API groups.
+* Follow API conventions: APIs exposed by every API server should adhere to [kubernetes API
+  conventions](../devel/api-conventions.md).
+* Support discovery API: Each API server should support the kubernetes discovery API
+  (list the suported groupVersions at `/apis` and list the supported resources
+  at `/apis/<groupVersion>/`)
+* No bootstrap problem: The core kubernetes server should not depend on any
+  other federated server to come up. Other servers can only depend on the core
+  kubernetes server.
+
+## Implementation Details
+
+### Reverse proxy
+
+A very basic Go program will suffice as a reverse proxy. Each federated
+API server will register its baseURL and swagger spec path with the proxy. This reverse proxy
+will summarize the list of all group versions exposed by all registered API
+servers at `/apis`.
+The Go program can also be replaced by any of the multiple open source API management tools
+(for example, there is [Kong](https://getkong.org/), which is written in lua and there is
+[Tyk](https://tyk.io/), which is written in Go).
+
+### Storage
+
+Each API server is responsible for storing their resources. They can have their
+own etcd or can use kubernetes server's etcd using [third party
+resources](../design/extending-api.md#adding-custom-resources-to-the-kubernetes-api-server).
+
+### Health check
+
+Kubernetes server's `/api/v1/componentstatuses` will continue to report status
+of master components that it depends on (scheduler and various controllers).
+Similiarly, proxy will expose a health check endpoint that will report the status of all
+federated api servers in the cluster.
+
+### Auth
+
+Since, the actual server which serves client's request is opaque to the client,
+all API servers need to have homogeneous authentication and authorisation mechanisms.
```

Comment:
Add: Currently, we'll be trusting system admins to configure all of the apiservers the same way.